core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/NetworkSenderActor.scala

```diff
@@ -1,23 +1,38 @@
 package edu.uci.ics.amber.engine.architecture.messaginglayer
 
-import akka.actor.{Actor, ActorRef}
-import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkOutputGate.{
+import akka.actor.{Actor, ActorRef, Props, Stash}
+import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkSenderActor.{
+  NetworkAck,
   NetworkMessage,
   QueryActorRef,
-  ReplyActorRef
+  RegisterActorRef,
+  SendRequest
 }
 import edu.uci.ics.amber.engine.common.ambermessage.neo.WorkflowMessage
 import edu.uci.ics.amber.engine.common.ambertag.neo.VirtualIdentity
 import edu.uci.ics.amber.engine.common.ambertag.neo.VirtualIdentity.ActorVirtualIdentity
 
 import scala.collection.mutable
 
-object NetworkOutputGate {
+object NetworkSenderActor {
+
+  /** to distinguish between main actor self ref and
+    * network sender actor
+    * TODO: remove this after using Akka Typed APIs
+    * @param ref
+    */
+  case class NetworkSenderActorRef(ref: ActorRef) {
+    def !(message: Any)(implicit sender: ActorRef = Actor.noSender): Unit = {
```

Comment:
I am not getting what is meant by this function. Are you overloading the ! operator?

Reply:
yes